fdbcli/LocationMetadataCommand.actor.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+ * LocationMetadataCommand.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbcli/fdbcli.actor.h"
+#include "fdbclient/Audit.h"
+#include "fdbclient/AuditUtils.actor.h"
+#include "fdbclient/IClientApi.h"
+#include "flow/Arena.h"
+#include "flow/FastRef.h"
+#include "flow/ThreadHelper.actor.h"
+
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+namespace {
+std::string describeUIDs(const std::vector<UID>& ids) {
+	std::string res;
+	for (const UID& id : ids) {
+		if (!res.empty()) {
+			res += ", ";
+		}
+		res += id.toString();
+	}
+
+	return res;
+}
+
+ACTOR Future<Void> printRandomShards(Database cx, int n, bool physicalShard) {
+	state Key begin = allKeys.begin;
+	state int numShards = 0;
+
+	while (begin < allKeys.end && numShards < n) {
+		// RYW to optimize re-reading the same key ranges
+		state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(cx);
+
+		loop {
+			try {
+				tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+				tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+
+				state RangeResult UIDtoTagMap = wait(tr->getRange(serverTagKeys, CLIENT_KNOBS->TOO_MANY));
+				ASSERT(!UIDtoTagMap.more && UIDtoTagMap.size() < CLIENT_KNOBS->TOO_MANY);
+
+				KeyRangeRef currentKeys(begin, allKeys.end);
+				RangeResult shards = wait(krmGetRanges(tr, keyServersPrefix, currentKeys, n, CLIENT_KNOBS->TOO_MANY));
+
+				for (int i = 0; i < shards.size() - 1 && numShards < n; ++i) {
+					const KeyRangeRef currentRange(shards[i].key, shards[i + 1].key);
+					std::vector<UID> src;
+					std::vector<UID> dest;
+					UID srcId;
+					UID destId;
+					decodeKeyServersValue(UIDtoTagMap, shards[i].value, src, dest, srcId, destId);
+					if (physicalShard == (srcId != anonymousShardId)) {
+						printf("Range: %s, IsPhysicalShard: %s, ShardID: %s, Src Servers: %s, Dest Servers: %s\n",
```

Comment:
Maybe too many shards to print? Do we want to stop early if sufficient amount of shards have been printed?

Reply:
The max number of shards is limited by `n`, provided by the user

Reply:
Yes. Thanks